Data/TLS/PThread/Internal.hs

```diff
@@ -73,35 +74,69 @@ delete k = do
 -- | A thread-local variable of type `a`.
 data TLS a = TLS { key       :: {-# UNPACK #-} !Key
                  , mknew     :: !(IO a)
-                 , allCopies :: {-# UNPACK #-} !(IORef [StablePtr a]) }
+                 , allCopies :: {-# UNPACK #-} !(IORef [Ptr (StablePtr a)])
+                 , destructor :: FunPtr (Ptr (StablePtr a) -> IO ())
+                 }
 
-mkTLS new = do
+mkTLS new = mkTLSWithDestructor new Nothing
+
+foreign import ccall "wrapper" wrapDestructor
+  :: (Ptr (StablePtr a) -> IO ()) -> IO (FunPtr (Ptr (StablePtr a) -> IO ()))
+
+-- | Like mkTLS but offers a destructor function which is called by any thread
+-- that created a copy of the variable when the thread exits. The destructor
+-- executes on a thread only if the thread exits before 'freeAllTLS' is called
+-- on the TLS variable.
+mkTLSWithDestructor new mdestructor = do
   evaluate check_error
-  key  <- easy_make_pthread_key
+  wrap <- maybe
+    (return nullFunPtr)
+    (\dfun -> wrapDestructor (peek >=> deRefStablePtr >=> dfun))
+    mdestructor
+  key  <- easy_make_pthread_key wrap
 --   putStrLn $ "KEY CREATED: "++show key
   allC <- newIORef []
-  return $! TLS key new allC
+  return $! TLS key new allC wrap
+
+-- Note [Ptr indirection]
+--
+-- getTLS check that the value of the pthread_key is not null
+-- to determine if a new copy of the TLS variable needs to be created.
+-- However, @castStablePtrToPtr <$> newStablePtr a@ might return @nullPtr@
+-- as the API of StablePtr does not guarantee that the resulting pointer is a
+-- valid address.
+--
+-- To ensure that getTLS does not create two copies for the same thread,
+-- we wrap the StablePtr in a malloc'ed Ptr and store that with setspecific.
+--
+-- Another undesirable consequence of confusing a StablePtr with nullPtr is
+-- that the destructor function of the pthread_key wouldn't be called in that
+-- case.
```

Comment:
This Note makes sense. However, I am worried about the performance implications of this extra indirection. Have you tried running the benchmarks in this package before and after this change?

(Also, do you know if `newStablePtr` returning `nullPtr` is actually an issue in practice?)